option java_package = "edu.washington.escience.myriad.proto";
option java_outer_classname = "LogicalRaProto";

message LogicalRaQueryMessage {
	required string name = 1;
	repeated LogicalRaOperator operators = 3;
}

message LogicalRaOperator {
	enum LogicalRaOperatorType {
		SCAN = 0;
		PROJECT = 1;
		EQUIJOIN = 2;
		CROSS = 3;
	}
	required LogicalRaOperatorType type=1;
	required string name = 2;

	optional LogicalScan scan = 3;
	optional LogicalSelect select = 4;
	optional LogicalProject project = 5;
	optional LogicalEquiJoin equijoin = 6;
	optional LogicalCross cross = 7;
}

message LogicalScan {
	required string relation = 1;
}

message LogicalSelect {
	required string childName = 1;
	required string condition = 2; // This must be improved!
}

message LogicalProject {
	required string childName = 1;
	repeated int32 columns = 2;
}

message LogicalEquiJoin {
	required string leftChildName = 1;
	repeated int32 leftColumns = 2;
	required string rightChildName = 3;
	repeated int32 rightColumns = 4;
}

message LogicalCross {
	required string leftChildName = 1;
	required string rightChildName = 2;
}
